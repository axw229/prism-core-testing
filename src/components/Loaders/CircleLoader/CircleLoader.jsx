// @flow
import React from 'react'

import './CircleLoader.scss'

type Props = {
  color?: string,
  className?: string
}

function CircleLoader (props: Props) {
  const { color, className, ...other } = props

  let circleProps = {}

  if (color) {
    circleProps = Object.assign({}, { stroke: color })
  }

  return (
    <svg className={`prism-loader-circle ${typeof className === 'string' ? className : ''}`} {...other} xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100' preserveAspectRatio='xMidYMid'>
      <circle cx='50' cy='50' {...circleProps} fill='none' strokeWidth='5' r='35' strokeDasharray='164.93361431346415 56.97787143782138' transform='rotate(143.836 50 50)'>
        <animateTransform attributeName='transform' type='rotate' calcMode='linear' values='0 50 50;360 50 50' keyTimes='0;1' dur='1s' begin='0s' repeatCount='indefinite' />
      </circle>
    </svg>
  )
}

export default React.memo<Props>(CircleLoader)
