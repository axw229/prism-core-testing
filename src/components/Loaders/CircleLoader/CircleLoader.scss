@import 'src/scss/convenience/_keyframe-animations.scss';

.prism-loader-circle {
  $dashDuration: 1.8s;
  $times: 5;
  $rotateDuration: $dashDuration / 2;
  $strokeColorDuration: ($dashDuration * 2) * $times;

  @keyframes strokeColor {
    0% {
      stroke: var(getVarName('loaders.circle.color1'));
    }
    20% {
      stroke: var(getVarName('loaders.circle.color2'));
    }
    40% {
      stroke: var(getVarName('loaders.circle.color3'));
    }
    60% {
      stroke: var(getVarName('loaders.circle.color4'));
    }
    80% {
      stroke: var(getVarName('loaders.circle.color5'));
    }
    100% {
      stroke: var(getVarName('loaders.circle.color1'));
    }
  }

  @keyframes dashAdjust {
    0% {
      stroke-dasharray: var(getVarName('loaders.circle.beginDash')) var(getVarName('loaders.circle.beginGap'));
    }
    100% {
      stroke-dasharray: var(getVarName('loaders.circle.endDash')) var(getVarName('loaders.circle.endGap'));
    }
  }

  animation: fadeOpacityIn 1s, rotate $rotateDuration infinite linear;
  display: block;
  height: auto;
  margin: auto;
  max-height: 90%;
  max-width: 90%;
  opacity: 1;
  stroke: currentColor;
  width: 3em;

  &__circle {
    animation: strokeColor $strokeColorDuration infinite cubic-bezier(0.075, 0.820, 0.165, 1.000), dashAdjust $dashDuration infinite alternate cubic-bezier(0.390, 0.575, 0.565, 1.000);
    stroke-linecap: round;
  }
}