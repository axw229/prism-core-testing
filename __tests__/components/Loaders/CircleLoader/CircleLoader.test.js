/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import CircleLoader from 'src/components/Loaders/CircleLoader/CircleLoader'

const loaderClass = 'color-wall-wall__loader'
// const loaderClass = 'prism-scene-manager__loader'
const color = 'red'

const getCircleLoader = (props) => {
  let defaultProps = {
    color: color,
    className: loaderClass
  }

  let newProps = Object.assign({}, defaultProps, props)
  return shallow(<CircleLoader {...newProps} />)
}

describe('CircleLoader with empty props', () => {
  let circleLoader
  beforeEach(() => {
    if (!circleLoader) {
      circleLoader = getCircleLoader({ color: '', className: '' })
    }
  })

  it('should match snapshot', () => {
    expect(circleLoader).toMatchSnapshot()
  })

  it('should render svg tag', () => {
    expect(circleLoader.find('svg').exists()).toBe(true)
  })

  it('should render circle tag', () => {
    expect(circleLoader.find('circle').exists()).toBe(true)
  })

  it('should render animateTransform tag', () => {
    expect(circleLoader.find('animateTransform').exists()).toBe(true)
  })
})

describe('CircleLoader with props', () => {
  let circleLoader
  beforeEach(() => {
    if (!circleLoader) {
      circleLoader = getCircleLoader()
    }
  })

  it('should match snapshot', () => {
    expect(circleLoader).toMatchSnapshot()
  })

  it('should render svg tag with class name defined as loaderClass constant', () => {
    expect(circleLoader.find('svg').hasClass(loaderClass)).toBe(true)
  })

  it('should render circle tag with prop stroke defined as color constant', () => {
    expect(circleLoader.find('circle').prop('stroke')).toEqual(color)
  })
})
