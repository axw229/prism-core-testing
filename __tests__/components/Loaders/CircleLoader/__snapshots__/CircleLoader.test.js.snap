// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleLoader with empty props should match snapshot 1`] = `
<svg
  className="prism-loader-circle "
  preserveAspectRatio="xMidYMid"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="50"
    cy="50"
    fill="none"
    r="35"
    strokeDasharray="164.93361431346415 56.97787143782138"
    strokeWidth="5"
    transform="rotate(143.836 50 50)"
  >
    <animateTransform
      attributeName="transform"
      begin="0s"
      calcMode="linear"
      dur="1s"
      keyTimes="0;1"
      repeatCount="indefinite"
      type="rotate"
      values="0 50 50;360 50 50"
    />
  </circle>
</svg>
`;

exports[`CircleLoader with props should match snapshot 1`] = `
<svg
  className="prism-loader-circle color-wall-wall__loader"
  preserveAspectRatio="xMidYMid"
  viewBox="0 0 100 100"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="50"
    cy="50"
    fill="none"
    r="35"
    stroke="red"
    strokeDasharray="164.93361431346415 56.97787143782138"
    strokeWidth="5"
    transform="rotate(143.836 50 50)"
  >
    <animateTransform
      attributeName="transform"
      begin="0s"
      calcMode="linear"
      dur="1s"
      keyTimes="0;1"
      repeatCount="indefinite"
      type="rotate"
      values="0 50 50;360 50 50"
    />
  </circle>
</svg>
`;
